### PR TITLE
graphql: metrics for validation and parsing phases

### LINKED ISSUE #3752: Metrics are missing for GraphQL validations
**Do you want to request a *feature* or report a *bug*?**

**What is the current behavior?**

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**

### PR DESCRIPTION
**Pull request by kamilkisiela**
New Prometheus metrics
- `query_parsing_time` 
- `query_validation_time`

Closes #3752

### CODE REVIEW

**Comment by leoyvens** [comment]
Seems there are build issues.

File: graph/src/data/query/result.rs [hunk 1]
```diff
@@ -45,12 +46,14 @@ pub type Data = Object;
 /// A collection of query results that is serialized as a single result.
 pub struct QueryResults {
     results: Vec<Arc<QueryResult>>,
+    pub validation_time: Option<Duration>,
```
Comment (leoyvens) [inline graph/src/data/query/result.rs:49]: Not great that this field had to be added. Could perhaps the GraphQlRunner have access to the `GraphQLServiceMetrics` and register the metric immediately after taking the measurement?

**Review by leoyvens — CHANGES_REQUESTED** [review]

File: graph/src/data/query/result.rs (hunk 1 above)
Comment (kamilkisiela) [reply graph/src/data/query/result.rs]: I created `GraphQLMetrics` that is passed to `GraphQlRunner` and is responsible for all metrics related to GraphQL:
- `query_execution_time`
- `query_parsing_time`
- `query_validation_time`
- `query_result_size`
- `query_result_max`

File: graph/src/components/graphql.rs [hunk 2]
```diff
@@ -43,6 +43,14 @@ pub trait GraphQlRunner: Send + Sync + 'static {
...
 
     fn load_manager(&self) -> Arc<LoadManager>;
+
+    fn metrics(&self) -> Arc<dyn GraphQLMetrics>;
+}
+
+pub trait GraphQLMetrics: Send + Sync + 'static {
+    fn observe_query_execution(&self, duration: Duration, results: &QueryResults) -> ();
+    fn observe_query_parsing(&self, duration: Duration, results: &QueryResults) -> ();
+    fn observe_query_validation(&self, duration: Duration, id: &DeploymentHash) -> ();
```
Comment (leoyvens) [inline graph/src/components/graphql.rs:53]: The `-> ()` can be omitted.

```suggestion
pub trait GraphQLMetrics: Send + Sync + 'static {
    fn observe_query_execution(&self, duration: Duration, results: &QueryResults);
    fn observe_query_parsing(&self, duration: Duration, results: &QueryResults);
    fn observe_query_validation(&self, duration: Duration, id: &DeploymentHash);
```

File: server/http/src/service.rs [hunk 3]
```diff
@@ -369,20 +315,26 @@ mod tests {
...
+
     lazy_static! {
         static ref USERS: DeploymentHash = DeploymentHash::new("users").unwrap();
     }
 
+    impl GraphQLMetrics for TestGraphQLMetrics {
+        fn observe_query_execution(&self, _duration: Duration, _results: &QueryResults) -> () {}
+        fn observe_query_parsing(&self, _duration: Duration, _results: &QueryResults) -> () {}
+        fn observe_query_validation(&self, _duration: Duration, _id: &DeploymentHash) -> () {}
+    }
```
Comment (leoyvens) [inline server/http/src/service.rs:336]: Same

File: server/http/tests/server.rs [hunk 4]
```diff
@@ -14,6 +14,14 @@ use graph_server_http::GraphQLServer as HyperGraphQLServer;
 
 use tokio::time::sleep;
 
+pub struct TestGraphQLMetrics;
+
+impl GraphQLMetrics for TestGraphQLMetrics {
+    fn observe_query_execution(&self, _duration: Duration, _results: &QueryResults) -> () {}
+    fn observe_query_parsing(&self, _duration: Duration, _results: &QueryResults) -> () {}
+    fn observe_query_validation(&self, _duration: Duration, _id: &DeploymentHash) -> () {}
```
Comment (leoyvens) [inline server/http/tests/server.rs:22]: Same

File: server/index-node/src/service.rs [hunk 5]
```diff
@@ -120,7 +120,7 @@ where
         let validated = ValidatedRequest::new(body, &req_parts.headers)?;
         let query = validated.query;
 
-        let query = match PreparedQuery::new(&self.logger, schema, None, query, None, 100) {
+        let query = match PreparedQuery::new(&self.logger, schema, None, query, None, 100, None) {
```
Comment (leoyvens) [inline server/index-node/src/service.rs:123]: I'd prefer to use a fake metrics collector like there is for tests, than have this be an option.

**Review by leoyvens — CHANGES_REQUESTED** [review]
Nice refactor!

File: graph/src/components/graphql.rs (hunk 2 above)
Comment (kamilkisiela) [reply graph/src/components/graphql.rs]: Ok, did

File: server/index-node/src/service.rs (hunk 5 above)
Comment (kamilkisiela) [reply server/index-node/src/service.rs]: Done

**Comment by kamilkisiela** [comment]
@leoyvens rebased

**Pull request merged**